// @nx/node/dist/src/generators/application/lib/add-dependencies.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.addProjectDependencies = addProjectDependencies;
const devkit_1 = require("@nx/devkit");
const internal_1 = require("@nx/js/internal");
const versions_1 = require("../../../utils/versions");
function addProjectDependencies(tree, options) {
    const bundlers = {
        webpack: {
            '@nx/webpack': versions_1.nxVersion,
        },
        esbuild: {
            '@nx/esbuild': versions_1.nxVersion,
            esbuild: internal_1.esbuildVersion,
        },
    };
    const exprPkgVersions = (0, versions_1.expressVersions)(tree);
    const koaPkgVersions = (0, versions_1.koaVersions)(tree);
    const fastifyPkgVersions = (0, versions_1.fastifyVersions)(tree);
    const frameworkDependencies = {
        express: {
            express: exprPkgVersions.expressVersion,
        },
        koa: {
            koa: koaPkgVersions.koaVersion,
        },
        fastify: {
            fastify: fastifyPkgVersions.fastifyVersion,
            'fastify-plugin': fastifyPkgVersions.fastifyPluginVersion,
            '@fastify/autoload': fastifyPkgVersions.fastifyAutoloadVersion,
            '@fastify/sensible': fastifyPkgVersions.fastifySensibleVersion,
        },
    };
    const frameworkDevDependencies = {
        express: {
            '@types/express': exprPkgVersions.expressTypingsVersion,
        },
        koa: {
            '@types/koa': koaPkgVersions.koaTypingsVersion,
        },
        fastify: {},
    };
    const typesNodeVersion = (0, versions_1.nodeTypesVersions)(tree).typesNodeVersion;
    return {
        installTask: (0, devkit_1.addDependenciesToPackageJson)(tree, {
            ...frameworkDependencies[options.framework],
            tslib: versions_1.tslibVersion,
        }, {
            ...frameworkDevDependencies[options.framework],
            ...bundlers[options.bundler],
            '@types/node': typesNodeVersion,
        }, undefined, options.keepExistingVersions ?? true),
        frameworkDependencies: frameworkDependencies[options.framework],
    };
}
